File: boj/2374.py
N=int(input())

chart = []

for i in range(N):
    chart.append(int(input()))

stack = [chart[0]]
maxi = chart[0]
cnt=0

# 올라 갈 때는 미리 더해준다는 느낌으로 더해주고
# 내려갈 때는 stack 을 교체만 하는 방식으로 한다.
# 마치고 stack에 잔류하고 있는 원소와 최대값의 차이를 더해주면 된다. 

for i in chart[1:]:
    if stack[0] > i:
        stack.pop()
        stack.append(i)
    elif stack[0] < i:
        cnt += (i - stack[-1])
        stack.pop()
        stack.append(i)
    else:
        pass
    maxi = max(maxi, i)

while stack:
    cnt += maxi - stack.pop()

print(cnt)

File: boj/2374.java
import java.io.*;
import java.math.*;
import java.util.*;

public class Main {

    static long[] arr;
    static Stack<Long> s;

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        int number = Integer.parseInt(br.readLine());
        arr = new long[number];
        s = new Stack<>();
        long ans = 0;
        long maxi;

        // 20 10 100 10 5 6 8
        for (int i = 0; i < number; i++)
            arr[i] = Integer.parseInt(br.readLine());
        maxi = arr[0];
        s.push(arr[0]);

        for (int i = 1; i < arr.length; i++) {
            if (s.peek() < arr[i]) {
                ans += arr[i] - s.peek();
            }
            s.pop();
            s.push(arr[i]);
            maxi = Math.max(arr[i], maxi);
        }

        ans += maxi - s.pop();

        System.out.println(ans);
        bw.close();
    }
}